### DISCUSSION TITLE
动态库链接的时候，缺少sanitizer 相关的 ldflags

### DISCUSSION THREAD

**Discussion by ivanallen** [xmake-io/xmake#4030] [Q&A]
### Xmake 版本

xmake v2.8.1+20230728

### 操作系统版本和架构

Linux 10-16-43-32 5.14.0-70.13.1.el9_0.x86_64 #1 SMP PREEMPT Wed May 25 21:01:57 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### 描述问题

使用 c++ 开发 python 的 so 模块，debug 版本会加上

```c++
if is_mode("debug") then
    add_cxflags("-fsanitize=address", "-fsanitize=undefined", "-fsanitize=leak", "-ftrapv")
    add_mxflags("-fsanitize=address", "-fsanitize=undefined", "-fsanitize=leak", "-ftrapv")
    add_ldflags("-fsanitize=address", "-fsanitize=undefined", "-fsanitize=leak", "-ftrapv")
end
```

编译普通的 binary 时，link 会加上 ldflags，但是编译 shared 共享库时不会。

这在开发 python 模块时会遇到问题，python 解释器通常没有去 link sanitizer，导致 python 在 load 对应的共享库时会报 undefined 错误。

![image](https://github.com/xmake-io/xmake/assets/12481610/95a8244c-4164-4f9e-b17c-3cfcc7c074ac)

### 期待的结果

link 动态库时也需要加上 -fsanitize=address 等相关的 link 选项。

### 工程配置

以下是一个最小复现的例子

```lua
add_rules("mode.debug", "mode.release")

if is_mode("debug") then
    add_cxflags("-fsanitize=address", "-fsanitize=undefined" )
    add_mxflags("-fsanitize=address", "-fsanitize=undefined")
    add_ldflags("-fsanitize=address", "-fsanitize=undefined")
end


target("xadd")
    set_kind("static")
    add_files("src/add.cc")
    add_headerfiles("src/add.h")

target("xadd-shared")
    set_kind("shared")
    add_files("src/add.cc")
    add_headerfiles("src/add.h")
    add_ldflags("-fsanitize=address")
    add_ldflags("-lasan")
```

![image](https://github.com/xmake-io/xmake/assets/12481610/e7c32ca9-02a4-45ac-a098-8a707f6e3b5e)

### 附加信息和错误日志

无

**Comment by ivanallen** [comment]
通过加 LD_PRELOAD 可解决。

sudo LD_PRELOAD=libasan.so.8:libubsan.so.1 python3 ../src/pymodule/test.py

[a comment by an automated account is not shown]

**Comment by waruqi** [comment]
shared 得用 add_shflags. 另外有内置的 `add_rules("mode.asan")` 直接能用 干嘛不用

```
xmake f -m asan
xmake
```

**Comment by ivanallen** [comment]
> shared 得用 add_shflags. 另外有内置的 `add_rules("mode.asan")` 直接能用 干嘛不用
> 
> ```
> xmake f -m asan
> xmake
> ```

我们想同时支持 asan ubsan 这些，这个配置是不是就不行？

**Comment by waruqi** [comment]
参考内部 rules 配置，自己定制

[a comment by an automated account is not shown]
